### ISSUE TITLE
|autocommit makes thousands of blank revisions

### ISSUE THREAD

**Issue by ohAitch** [urbit/urbit#3317]
**Describe the bug**

Turning on `|autocommit` steadily increments the clay revision, regardless of any change occurring.

**To Reproduce**

```hoon
> |autocommit
# echo >>zod/home/dojo.hoon
: /~zod/home/4/dojo/hoon
<<some time passes>>
# echo >>zod/home/dojo.hoon
: /~zod/home/1.049/dojo/hoon
```

**Expected behaviour**
The next revision after revision 4 being revision 5, and there being no interim NOP revisions.

**Screenshots**
See above, ish.

**System (please supply the following information, if relevant):**
 - OS: linux64 [busybox](https://github.com/asssaf/urbit-fakezod-docker)
 - Urbit v0.10.7 OS v1.0.25 😛
  (don't think the IPC rewrite actually touched this much?)
 - `%base` hash 0v1m.o19aa...cms4k, similarly dating back to the existence of /=base=

**Additional context**
N/A

**Notify maintainers**
@philipcmonk @joemfb

**Comment by ryjm** [comment]
i left `|autocommit` on for a while, same thing happened for me:
```
> |commit %home
>=
+ /~littel-wolfur/home/129.556/dojo/hoon
> |autocommit %home
>=
: /~littel-wolfur/home/129.826/dojo/hoon
: /~littel-wolfur/home/129.900/dojo/hoon
```

also i was bail:meme'ing every 4 hours or so (and fixing with `uniq`), i wonder if this could be the cause...

**Comment by ryjm** [comment]
yeah, if you do a bunch of `|autocommit`'s you'll see your total userspace memory go up by about 1MB a second. I think what happened is i ran `|autocommit %home` thinking it was canceling it, when it was really creating another instance of it.

**Comment by philipcmonk** [comment]
129k revisions, whew!  I could definitely see this creating a bunch of duplicate nouns if it's recreating the files every time.  It should check if the built file is equal to the previous revision and no-op if no files differ, guess it doesn't.

Notably, this also happens with `|commit`:

```
> -:.^(* %cw %)
2
> |commit %home
>=
> -:.^(* %cw %)
3
```

Separately, `|autocommit` is definitely a hack and I only ever run it on fake ships.  With Ford Fusion now I would consider it though (after this is fixed).

**Comment by ohAitch** [comment]
*something* is certainly checking whether they're equal to the previous
revision, since the modified files have change notifications printed.

**Comment by philipcmonk** [comment]
Indeed, which comes from:

https://github.com/urbit/urbit/blob/b0d252fa7633ba030c555f690594d85c0d59b36f/pkg/arvo/sys/vane/clay.hoon#L1458-L1482

This is pre-conversion data though, which means if a file is committed as a page rather than a lobe (eg if committed under a different mark than the filename dictates, which means it will be converted), then it will be listed as a change, even if it wasn't.  E.g:

```
> *%/test/hoon 'hi'
+ /~zod/home/5/test/hoon
> *%/test/hoon 'hi'
: /~zod/home/6/test/hoon
```

Notably this does not happen with `|commit` because it gets filtered against our mime cache very early so we don't process large commits every time we restart urbit:

https://github.com/urbit/urbit/blob/b0d252fa7633ba030c555f690594d85c0d59b36f/pkg/arvo/sys/vane/clay.hoon#L1332-L1353

This suggests three answers:

- At the end of +park (unified commit logic), if the generated (map path lobe) is equal to the old one, revert all changes (ie save `+>.$` at the beginning and restore it).
- If get-changes gives no changes, no-op with no commit (accepting the false positive noted above but avoiding empty `|(auto)commit`).  Ie use the same heuristic as printing does now.
- No-op in +into if empty

I think the first is best, though the second is slightly easier and correct most of the time.

**Comment by crides** [comment]
I built and installed urbit on `release/next-sys`, which includes #3376, but it seems like the problem has not been fixed. I still get intermitent automatic revision increments (which also allocates RAM at around 1.5MB/sec which is really visible in `htop`), but I also run multiple `|autocommit`s because a lot of times `|autocommit` just doesn't work. How can I help debug this?

**Comment by philipcmonk** [comment]
Hmm, I just booted a ship off master, now that `release/next-sys` has been added, and I can see that `|autocommit` is working (ie running every couple seconds) but the revision number is not increasing (per `.^(* %cw %)`).

Did you use `-B ~/urbit/bin/solid.pill` so it uses the pill on the branch you're on?  The public one is a couple weeks old I think, so it wouldn't have it.  Also I don't know if a new pill was created after #3376.
